compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1009,20 +985,83 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 .or_insert(binding.span);
         }
 
+        // Include substitutions for generic parameters of associated types
+        let candidate = candidate.map_bound(|trait_ref| {
+            let item_segment = hir::PathSegment {
+                ident: assoc_ty.ident,
+                hir_id: None,
+                res: None,
+                args: Some(binding.gen_args),
+                infer_args: false,
+            };
+
+            let substs_assoc_item = self.create_substs_for_associated_item(
+                tcx,
+                path_span,
+                assoc_ty.def_id,
+                &item_segment,
+                trait_ref.substs,
+            );
+
+            debug!(
+                "add_predicates_for_ast_type_binding: substs for assoc_item: {:?}",
+                substs_assoc_item
+            );
+            ty::TraitRef::new(trait_ref.def_id, substs_assoc_item)
```

Comment:
This trait ref is kind of...not correct. We're still using the trait `DefId`, so there are some extra substs at the end. But...`ProjectionTy::from_ref_and_name` (below) expects it like this.

Instead, can you make this use the associated ty def id? And then you also don't need to use `from_ref_and_name`.

Reply:
Maybe a poor choice for the variable name, but `substs_assoc_item` does contain the substs for the trait ref:

As an example for:

```
#![feature(generic_associated_types)]

trait X<'a, T> {
  type Y<'b>;

  fn bla(&self, &'a T) {}
}

struct Foo {}

impl Foo {
  fn foo<'a, T>(&self, arg : Box<dyn for<'b> X<'a, T, Y<'b> = &'b ()>>) -> u32 {
    5   
  }
}

fn main() {}
```

we get back the following for the debug statement [here](https://github.com/rust-lang/rust/blob/f2cb963b7a117446e7f5ca8f8a7a66813113342e/compiler/rustc_typeck/src/astconv/mod.rs#L1006):

```DEBUG rustc_typeck::astconv add_predicates_for_ast_type_binding: substs for assoc_item: [FreshTy(0), ReLateBound(DebruijnIndex(1), BrNamed(DefId(0:12 ~ gat_trait_path_playground[317d]::{impl#0}::foo::'a), 'a)), T, ReLateBound(DebruijnIndex(0), BrNamed(DefId(0:14 ~ gat_trait_path_playground[317d]::{impl#0}::foo::'b), 'b))]```

Is this incorrect?

Reply:
Yes, you're right. The substs for the associated type include the substs for the parent item. However, the problem for this line is that the trait ref is *not* for the associated type, it's for the parent trait (`trait_ref.def_id`). So instead of `[FreshTy(0), 'a, T]` we have `[FreshTy(0), 'a, T, 'b]`.

Now, this doesn't exactly end up being a problem, since we don't use the trait ref defid between here and `from_ref_and_name`, and this is actually the form that `from_ref_and_name` expects. However, `from_ref_and_name` was probably created pre-GATs, where the substs for an associated type will always equal the substs of the parent trait ref. And it's a bit scary and confusing that this trait ref gets created like this.

If you just use the associated item `DefId`, then this is completely fine, since the substs we have match the substs we expect. But, you would have to change the `from_ref_and_name` call below, since that expected the `def_id` to be the parent trait DefId, not the associated item's. (And this is actually a bit nicer because we already have the associated item's `DefId` and don't have to look it up, which happens in `from_ref_and_name`.

Reply:
Thanks for the answer. Mabye I misunderstand you here, do you want to use the `DefId` of the `AssocItem` as the `DefId` in `ty::TraitRef`? This also doesn't seem correct. I mean we could add a comment to warn that this isn't the correct `DefId`, but then we could also add a comment  to warn that this trait ref doesn't contain the right substs in the current commit. Seems like both options aren't optimal, but if you prefer the `DefId` change I will change that.

Reply:
What if we change `ty::AssocItem` to contain `SubstsRef` (shouldn't this exist for GATs anyway?) and then create a `ty::AssocItem` here. Then create a function that creates a `ProjectionTy` from an AssocItem given substs for it's parent and use that function instead of `ProjectionTy::from_ref_and_name`?

Reply:
> Thanks for the answer. Mabye I misunderstand you here, do you want to use the `DefId` of the `AssocItem` as the `DefId` in `ty::TraitRef`? This also doesn't seem correct. I mean we could add a comment to warn that this isn't the correct `DefId`, but then we could also add a comment to warn that this trait ref doesn't contain the right substs in the current commit. Seems like both options aren't optimal, but if you prefer the `DefId` change I will change that.

Hmm, I suppose you're right: a `TraitRef` here doesn't make much sense. Instead, let's change this to be
```
let assoc_substs = candidate.map_bound(|trait_ref| {
   ...
   substs_assoc_item
});
```
where then `assoc_substs: Binder<SubstsRef<'tcx>>`.

Then most of what I said otherwise still holds. Except, instead of doing `candidate.map_bound` below, you would do `assoc_substs.map_bound`.

Alternatively, you could make this block return an `projection_ty: Binder<ProjectionTy<'tcx>>` by returning `ProjectionTy { item_def_id: assoc_ty.def_id, substs: substs_assoc_item }`. Then, in the `Equality` block below, you can do
```
projection_ty.map_bound(|projection_ty| {
  ty::ProjectionPredicate { projection_ty, ty }
})
```
and in the `Constraint` block
```
let param_ty = tcx.mk_ty(Projection(projection_ty.skip_binder()));
...
```

I actually prefer the second option, because it makes it more clear what's happening.

Reply:
> What if we change `ty::AssocItem` to contain `SubstsRef` (shouldn't this exist for GATs anyway?) and then create a `ty::AssocItem` here. Then create a function that creates a `ProjectionTy` from an AssocItem given substs for it's parent and call that function where `ProjectionTy::from_ref_and_name` is currently used?

Let's hold off on changing other things for now, since I have to think through those more.

Reply:
Don't we need the `TraitRef` in `collect_constrained_late_bound_regions` when we check [here](https://github.com/rust-lang/rust/blob/a40f7ecb291f3d0ccb47658f2a6044ca83a82380/compiler/rustc_typeck/src/astconv/mod.rs#L1014-L1048) whether the `Ty` of the binding contains any lifetimes not mentioned in either the trait ref or the associated type?

Reply:
Nope that only needs a `TypeFoldable`, which `SubstsRef` is. (Technically in a `Binder`, but we have that here)